docs/man/stepshrink.1
.TH stepshrink 1 "January 16th, 2003"
.SH NAME
stepshrink - reduce the size of pyDDR step files
.SH SYNOPSIS
\fBstepshrink \fR[\fB\-o \fInewfilename.step\fR] \fIfilename.step\fR
.SH DESCRIPTION
Most step files contain a lot of redundancy; rows of empty notes, eighth
notes where there could be half notes, and so on. Stepshrink examines the
steps present in a file, and reduces the size by using more appropriate
directives in the file. No information in the file is lost doing this.
.PP
Stepshrink also reduces the accuracy of lyrics timings to the nearest
1/100th of a second, rather than the hyperaccurate 16 digits that is the
default. This does lose information, but not noticably.
.PP
By default, stepshrink outputs to standard output (your terminal), but
the \-o command line option can make it output to a file. It is safe to
specify the same file as both the input and output file.
.SH AUTHOR
Joe Wreschnig <piman@debian.org> is the author of stepshrink, the
maintainer of pyDDR for Debian, and the author of this manual page.
.SH SEE ALSO
\fBpyddr\fR(6)


ChangeLog
pydance ChangeLog

legend:
+ new feature added or made an existing one better
- removed a feature, or something bad happened
* fixed a bug

----- - CURRENT
0.7.5 - That's the second largest ChangeLog I've ever seen!	[2003.XX.XX]
+ New options: "Wide" to add jumps, holds on/off.
+ 5 panel (Pump It Up) support (single, versus/unison, couple, and double).
 + Support for KSF, SM, and .dance files with 5 panel steps.
 - Actual PIU pads may or may not work; none were available for testing.
+ Versus, couple, and double 6 and 8 panel modes.
+ New game type selector, so you can do e.g. Double/Endless.
+ A document on the KSF file format has been added in docs/.
+ The .dance specification has been updated with new modes and corrections.
+ A new centered display for single player modes.
+ New theming code.
 + Brings back oldschool and bryan themes.
 + Autodetect what game modes a theme will work for.
+ New 32x32 arrow set for 8/9 panel modes, by xsdg.
* Support the Thrustmaster USB mats (probably), and X-Box controllers.
 + We suspect that pydance will run on GNU/Linux on an X-Box now.
* Correctly encode/decode KSF, SM, DWI, and .dance file metadata.
 * KSF requires the Python Koco codecs package.
* DWI parsing for braindead constructs like "2!(2".
* Arrows now always draw in order (earlier under later).
* Many crashes fixed:
 * Stops + holds works again.
 * Oni lifebar works again.
 * Obscure edge case in 2 player endless mode works correctly.
* Parsing for all files is now more informative and less error-prone.
- Assist mode is broken for non-4 panel modes, we need more sounds.
- any2dance removed, it won't work anymore with this version.
- In-game menu help removed, due to inaccuracies.

0.7.4 - In My Pants						[2003.07.28]
+ 6 panel ("solo") and 8 panel support.
+ Double (1 player, both sides) support.
 - 'totaljudgings' option no longer works (no one was using it anyway...)
* Song selector UI tweaks (faster animations, nicer folder opening).
* Respace the arrows to be closer together in single/versus.
* All 9 directions can always be done on the numeric pad for player 1.
 - azerty 2nd player start key changed to '+' on the numeric pad.
* "Escape" changed from left+start to start+select.
* Some more speed optimizations for slower processors.
* Some DWI parser fixes for bad files.
* 'usepsyco' option (on by default) to try and load the Psyco JIT compiler.
- 'oldschool' and 'bryan' were removed (temporarily, maybe).

0.7.3
* Skipped, because the Linux Game Tome called 0.7.2 0.7.3 by mistake.

0.7.2 -	Dilaudid, how I love thee                               [2003.07.14]
+ Boost, brake, left, right, mirror, shuffle, and random ("supershuffle").
+ "Suck" lifebar mode, you fail if you do too good.
+ "Drop" lifebar mode, goes down at the regular speed but never up.
+ Folders in the song selector (needs > 60 songs).
* Parser bug fixes (especially for SM files).
* pyDDR has been renamed to pydance.
	
0.7.1 - OUCH MY BLADDER                                         [2003.07.01]
+ Support for COUPLE mode.
 + And battle mode.
 - any2dance doesn't always convert COUPLE from DWI to .dance correctly.
+ New player options screen (hold down start in the song selector).
+ New game options screen (hold down select).
+ "Ready? Go!" screen before the song starts.
+ New "oldschool" theme similar to earlier mixes.
* Standard round of bug fixes and tweaks.

0.7.0 - Applesauce makes the apples go yonder                   [2003.06.04]
+ More optimizations from JZig (movies this time)
+ New file formats
  + Native DWI and SM support
  + ".dance" instead of ".step" is now pyDDR's official format.
  + 'any2dance.py' converter for files.
* Judge rewrite!
  * Proper speed change handling in Judge
  * Proper miss handling in Judge

0.6.5 - Boy, did I mess that up.				[2003.05.30]
+ Major graphics optimization patch (1.5x-2x increase in fps) from JZig.
* Two player mode works again.

0.6.4 - RTFM. RTFM. RTFM. That means Read the Fucking Manual.	[2003.05.29]
+ New step file parsers
  + Beginner and hardcore (smaniac/oni/challenge) difficulty.
+ New hold arrow display
- 'mixerclock' option removed (it's now the only timing method).
- 'transcolor' option removed in favor of new 'lyriccolor' syntax.
* Bug fixes, optimizations, whippings, etc.
  + Most notably, holds, bpm changes, and stops are working much nicer now.
    - But may require reconversion with dwi2step.

0.6.3 - Reasonable Starting Times!				[2003.05.17]
+ New song loading display and algorithm (at least 4-5 times faster).
+ New song selector based roughly on DDR 5th Mix.
  + Nonstop help and playlist display.
  + 'gratuitous' option to turn off the slow extras in the song selector.
+ Endless mode.
+ Configurable oni life meter.
+ New "centered" arrow scrolling style for gameplay.
+ Smaller download size by converting wavs to oggs.
  - Requires SDL_mixer 1.2.5.
+ The timer display is controllable via the 'fpsdisplay' option.
+ New theme ('blue') and a new life display for all themes.
* DDR Extreme (8th Mix) grading and scoring algorithm.
* Split off the ddrmat.o driver from the main package.
* The constant cleanup/speedup/refactor/bug fix process.

0.6.2 - No More Ugly-Ass Buttons				[2003.04.28]
- Fullscreen switching during the game doesn't work on MacOS X.
  - Use the configuration file instead, for now.
- The pN_keys option is gone in favor of 'keyboard', and the 'sortpersist'
  option was removed.
+ 'killsongonfail' was renamed to 'autofail' and made default.
+ New game menu display.
+ Better theme detection and configuration support.
+ The 'badknees' option will remove jumps.
+ A new README that actually documents all the options!
* Once again, numerous cleanups and bugs fixed.

0.6.1 - The second one						[2003.04.09]
+ Experimental install system
  + Currently only works on Windows (and doesn't do much) and UNIX.
  + Manual pages included with pyDDR now.
  + New directory layout
+ Major code cleanups and bug fixes.

utils/stepshrink.py
#!/usr/bin/python
# Shrink the size of PyDDR step files
# By Joe Wreschnig <piman@debian.org>

import sys
import re
import getopt

VERSION = "0.9"

def usage():
  print "stepshrink " + VERSION + " - shrink PyDDR step files"
  print "Usage: " + sys.argv[0] + " [-o output] stepfile"
  print "  --output, -o    Filename to output to (default stdout)"


def main():

  output_file = None

  try:
    opts, args = getopt.getopt(sys.argv[1:], "o:vh",
                               ["output", "version", "help"])
  except getopt.GetoptError:
    usage()
    sys.exit(2)

  for o, a in opts:
    if o in ("--help", "-h"):
      usage()
      sys.exit()
    elif o in ("-v", "--version"):
      print "stepshrink version " + VERSION
      sys.exit()
    elif o in ("-o", "--output"): output_file = a

  if len(args) == 0:
    usage()
    sys.exit(2)

  outfile = sys.stdout
  lines = []
  lines2 = []

  infile = open(args[0])
  lines = infile.readlines()
  infile.close()

  lines = lyrics_and_delay(lines)

  types = ("sixty", "thrty", "steps", "eight", "qurtr", "halfn", "whole")
  for i in range(0,len(types) - 1):
    lines = remove_empty_types(types[i], types[i+1], lines)

  lines = remove_empty_types("twtfr", "tripl", lines)

  lines = delays_to_wholes(lines)
  lines = consecutive_delays(lines)

  if output_file != None: outfile = open(output_file, "w")
  for line in lines: outfile.write(line + "\n")
  if output_file != None: outfile.close()

def lyrics_and_delay(lines):
  e_eight = re.compile("^\s*eight\s+00\s+00\s+00\s+00")
  atsec = re.compile("^\s*atsec\s+(\d+)\.(\d*)")
  version = re.compile("^\s*version\s+\d+.\d+")
  btm = re.compile("^\s*[BTM]")
  lyr = re.compile("^\s*L")
  i = 0

  lines2 = []
  stage = 0
  for line in lines:
    if stage == 0:
      if version.match(line):
        stage = 3
        line = "version 0.6" # thrty support requires this
      lines2.append(line.strip())
      
    elif stage == 1:
      # Incredible precision isn't really necessary for lyrics...
      if line == "end": stage = 3
      res = atsec.match(line)
      if res: line = "atsec %s.%s" % (res.group(1), res.group(2)[0:2])
      lines2.append(line.strip())

    elif stage == 2:
      if not e_eight.match(line):
        if i % 2 == 1: lines2.append("eight 00 00 00 00")
        if i > 1: lines2.append("delay %d" % (i/2))
        lines2.append(line.strip())
        stage = 3
        i = 0
      else:
        i += 1

    else:
      if btm.match(line): stage = 2
      elif lyr.match(line): stage = 1
      lines2.append(line.strip())

  return lines2

def remove_empty_types(type, double_type, lines):
  lines2 = []
  emp = re.compile("^\s*%s\s+00\s+00\s+00\s+00" % type)
  ne = re.compile("^\s*%s\s+(\d\d)\s+(\d\d)\s+(\d\d)\s+(\d\d)" % type)
  for line in lines:
    if emp.match(line) and ne.match(lines2[-1]):
      res = ne.match(lines2[-1])
      lines2[-1] = "%s %s %s %s %s" % (double_type, res.group(1),
                                       res.group(2), res.group(3),
                                       res.group(4))
    else: lines2.append(line)
  return lines2

def delays_to_wholes(lines):
  lines2 = []
  delay = re.compile("^\s*delay\s+(\d+)")
  qurtr = re.compile("^\s*qurtr\s+(\d\d)\s+(\d\d)\s+(\d\d)\s+(\d\d)")

  for line in lines:
    res = delay.match(line)
    if len(lines2) != 0: res2 = qurtr.match(lines2[-1])
    else: res2 = None
    if res and res2 and int(res.group(1)) > 0:
      lines2.pop()
      i = int(res.group(1))
      if i >= 3:
        lines2.append("whole %s %s %s %s" % res2.groups((1,2,3,4)))
        lines2.append("delay %d" % (i - 3))
      elif i >= 1:
        lines2.append("halfn %s %s %s %s" % res2.groups((1,2,3,4)))
        lines2.append("delay %d" % (i - 1))
    else: lines2.append(line)
  return lines2

def consecutive_delays(lines):
  lines2 = []
  delay = re.compile("^\s*delay\s+(\d+)")

  for line in lines:
    res = delay.match(line)
    if len(lines2) != 0: res2 = delay.match(lines2[-1])
    if res and res2:
      i = int(res.group(1))
      j = int(res.group(2))
      if i + j == 0: lines2.pop()
      else: lines2[-1] = "delay %d" % (i + j)
    elif not res or res.group(1) != "0": lines2.append(line)
  return lines2

if __name__ == '__main__':
  main()
